File: sprint02/JMS/PGS_43162.cpp
#include <string>
#include <vector>

using namespace std;

int arr[200][200]={0};
int visited[200]={0};

void DFS(int n, int vertex){
    visited[vertex] = 1;
    
    for(int i=0;i<n;i++){
        if(arr[vertex][i] && !visited[i])
            DFS(n,i);
    }
}

int solution(int n, vector<vector<int>> computers) {
    int answer = 0, i, j;
    
    vector<vector<int>>::iterator comit;
    for(comit = computers.begin(), i=0; comit != computers.end() ; comit++, i++){
        vector<int>::iterator it;
        for(it = (*comit).begin(), j=0; it != (*comit).end(); it++, j++){
            if(*it)
                arr[i][j]=1;
        }
    }
    
    for(int i=0;i<n;i++){
        if(!visited[i]){
            answer++;
            DFS(n,i);
        }
    }
    
    return answer;
}

File: sprint02/KSH/PGS_43162.py
def dfs(idx, dic, visited):
    if idx in visited:
        return
    visited.append(idx)

    for i in dic[idx]:
        dfs(i, dic, visited)
        
# def dfs(idx, dic, visited):
#     for i in dic[idx]:
#         if i not in visited:
#             visited.append(i)
#             dfs(i, dic, visited)

def solution(n, computers):
    answer = 0
    
    dic = dict()
    for i in range(n):
        dic[i + 1] = []
    for idx, i in enumerate(computers, start=1):
        for idx_, j in enumerate(i, start=1):
            if idx == idx_:
                continue
            if j == 1:
                dic[idx].append(idx_)
                
    visited = []
    
    for i in range(1, n + 1):
        if i not in visited:
            answer += 1
            dfs(i, dic, visited)
    
    return answer